Refactor creation of bootstrap DBs

mrtracy opened the pull request:
This commit refactors the way that the SystemDB and its associated tables are
created, in order to allow for additional arbitrary tables to be created on a
cluster which is being bootstrapped.

Commit creates a new "MetadataSchema" to which databases and tables can be
added; MetadataSchema exposes a "GetInitialValues()" method which returns the
necessary KeyValue objects to bootstrap the schema onto a new Cockroach cluster.

The descriptors for SystemDB tables are still created statically, as some of
them (such as DescriptorTable) are needed during the bootstrap process. These
descriptors are added directly to the MetadataSchema.

Non-System tables can also be added; in the case of these tables, the SQL schema
definition is provided to SystemDB rather than a descriptor.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/3322)

<!-- Reviewable:end -->

Commit:
Refactor creation of bootstrap DBs

This commit refactors the way that the SystemDB and its associated tables are
created, in order to allow for additional arbitrary tables to be created on a
cluster which is being bootstrapped.

Commit creates a new "MetadataSchema" to which databases and tables can be
added; MetadataSchema exposes a "GetInitialValues()" method which returns the
necessary KeyValue objects to bootstrap the schema onto a new Cockroach cluster.

The descriptors for SystemDB tables are still created statically, as some of
them (such as DescriptorTable) are needed during the bootstrap process. These
descriptors are added directly to the MetadataSchema.

Non-System tables can also be added; in the case of these tables, the SQL schema
definition is provided to SystemDB rather than a descriptor.

petermattis commented:
LGTM

---

Review status: 0 of 11 files reviewed at latest revision, 2 unresolved discussions.

---

*[sql/metadata.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4hKe2UHL9udK5bHwKZ:-K4hKe2VsHt3PzGKOtbx:-1612158754) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L157)):*
Dangling comment.

---

*[storage/client_test.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4hXhOcQZWOVYpFIJeV:-K4hXhOcQZWOVYpFIJeW:-837890296) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/storage/client_test.go#L130)):*
Would we ever create a `MetadataSchema` without adding the system database to it?  Perhaps the `AddSystemDatabaseToSchema` call should be moved into `NewMetadataSchema`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-:-K4hZVyVt4UAOF4beQoS:676834233)*
<!-- Sent from Reviewable.io -->

tschottdorf commented on sql/metadata.go:
```diff
@@ -0,0 +1,160 @@
+		Value: value,
+	})
+
+	// Sort returned keys; this
```
this..?

tschottdorf commented on sql/metadata.go:
```diff
@@ -0,0 +1,160 @@
+
+// AddSystemDescriptor adds a new system descriptor to the schema. System
+// descriptors have well-known, static descriptors; however, the MetadataSchema
+// is used to generate the KeyValue objects needed install them on a new
```
Needed to

tschottdorf commented on sql/system.go:
```diff
@@ -103,91 +102,55 @@ var (
+	// above (SystemDB and each system table). This is used by tests which need
+	// to know the number of system descriptors being installed; if you add a
+	// new system table, increment this number.
+	NumSystemDescriptors = 6
```
Can't be automated?

tamird commented on sql/metadata.go:
```diff
@@ -0,0 +1,160 @@
+
+// NewMetadataSchema constructs a new MetadataSchema object which contains no
+// tables.
+func NewMetadataSchema() *MetadataSchema {
```
why does this need to return a pointer? actually, why bother with a constructor at all?

tamird commented on sql/metadata.go:
```diff
@@ -0,0 +1,160 @@
+type MetadataDatabase struct {
+	name       string
+	privileges *PrivilegeDescriptor
+	tables     []*metadataTable
```
why pointers?

mrtracy commented:
Review status: 0 of 11 files reviewed at latest revision, 7 unresolved discussions.

---

*[sql/metadata.go, line 48 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-48:-K4xihu674CXKYoo9LBo:-1551113873) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L48)):*
See below.

---

*[sql/metadata.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-58:-K4xiT_4BNiAI0xvr_VK:1021409186) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L58)):*
I've refactored this system so that there are no more pointer returns.  This is now MakeMetadataSchema(); I kept it because I also accepted Peter's suggestion of always including the system database in every new MetadataSchema, so this method is now non-trivial.

---

*[sql/metadata.go, line 64 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-64:-K4xMKpG54VW7UjDz2I7:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L64)):*
Done.

---

*[sql/metadata.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-157:-K4xMJYolOlW32yLbRXb:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L157)):*
Done.

---

*[sql/metadata.go, line 157 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4hKe2UHL9udK5bHwKZ:-K4xMJvm21w0ejzK1OXO:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L157)):*
Done.

---

*[sql/system.go, line 109 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMq-r1-109:-K4xijtMo9PYYi8QiIuf:1486013123) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/system.go#L109)):*
I misunderstood how this was being automated before; I have restored the automation and updated the comment appropriately.

---

*[storage/client_test.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4hXhOcQZWOVYpFIJeV:-K4xLCIv_izNG9veuOjx:-188318272) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/storage/client_test.go#L130)):*
I was establishing the pattern for usage by other packages; for instance, this is the pattern the 'storage' package should use to do this.  However, I think you are correct, there's not much of a use case for separating the system database like this at this time.  Updated.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 11 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/metadata.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-58:-K4xjQNTcGAXPyYORctS:-1773361051) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L58)):*
Oh, okay, so this new code you have here does not do what you want (I think).

You're creating `ms`, taking a reference and passing to `AddSystemDatabaseToSchema`, and then returning a COPY of that thing. Either revert this to pointers, or (if pointer equality doesn't matter), change the argument to `AddSystemDatabaseToSchema` to by-value, rather than by-reference.

---

*[sql/metadata.go, line 16 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4xkHNXBKa0KrgHsSOe:-K4xkHNXBKa0KrgHsSOf:-11500317) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a57e958f4da6c30432205efb100044a26142c/sql/metadata.go#L16)):*
wrong author

---

*[sql/system_test.go, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4xkQoP97DHqFt3SvUV:-K4xkQoP97DHqFt3SvUW:-1877162651) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a57e958f4da6c30432205efb100044a26142c/sql/system_test.go#L35)):*
what's going on here and on the next line?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 11 files reviewed at latest revision, 8 unresolved discussions.

---

*[sql/metadata.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-58:-K4xmNROAOOJyZvoIR2M:-1233831325) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L58)):*
There's a test that explicitly demonstrates that this does what I want: returns a MetadataSchema which has the default system tables in it.  Why wouldn't this work?  The value returned definitely has the correct objects in it.   AddDatabase() and AddTable() still have pointer receivers.

---

*[sql/metadata.go, line 16 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4xkHNXBKa0KrgHsSOe:-K4xmgjLckDkkF0QMea5:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a57e958f4da6c30432205efb100044a26142c/sql/metadata.go#L16)):*
Done.

---

*[sql/system_test.go, line 35 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4xkQoP97DHqFt3SvUV:-K4xn5mbIpvksLKzuZ-g:-1435529777) ([raw file](https://github.com/cockroachdb/cockroach/blob/c45a57e958f4da6c30432205efb100044a26142c/sql/system_test.go#L35)):*
Only the first kv := ms.GetInitialValues() is needed, just a duplicated statement.  Fixed.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/metadata.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-58:-K4xp8LCBYHEBOK1bIFm:-1218388896) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L58)):*
Ah, `AddSystemDatabaseToSchema` just modifies the receiver. OK, this is fine.

It seems that `AddSystemDatabaseToSchema` is only called here, though, so maybe it should be inlined here?

If not, make it a method to clarify that it modifies `ms` (and not global state, which is what I suspected), and unexport it.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
Review status: 9 of 11 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/metadata.go, line 58 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322#-K4eQE4TjL2BME6yfiMu-r1-58:-K4y8E133bCQgt3b18F4:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/80eb5b970021e09c4b943cc7e5feeb96eaac24fe/sql/metadata.go#L58)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r4.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3322)*
<!-- Sent from Reviewable.io -->

mrtracy commented:
@tamird I'm assuming this is LGTM?

tamird commented:
yep
